### ISSUE TITLE
Add `after_instantiate_classes` hook to LightningCLI to save LightningDataModule information in the current log directory

### ISSUE THREAD

**Issue by tchesler** [Lightning-AI/lightning#18608]
### Description & Motivation

Hi. 

I noticed that `LightningCLI` has a member function called `before_instantiate_classes` called before `instantiate_classes` but no `after_instantiate_classes`. I need to save the data split generated in `LightningDataModule` after it has been instantiated to a file in the current logging directory. And this is only possible after `Trainer.instantiate_classes`, because only then I have access to `trainer.logger.log_dir` and `LightningCLI.datamodule` in order to save files related to the current run in the current logging directory.

I tried to do it with a custom Callback for the Trainer overwriting `on_fit_start()`, however, the trainer only has access to the dataloaders, and not the underlying datamodule so you can't do it that way.

In my specific usecase, I want to save and load the training/validation/test data split to and from a separate file based on the result of `torch.utils.data.random_split` or if the filepath are given, based on the given files. The goal is to train different models (in different LightningCLI instances) on the same data split given the same underlying data, without doing requiring setting any seed, because there can be a case where you only pick some of the data to run validation on, and not all of them.
```python
# In the LightningDataModule
data = #Inherits for example from torch DatasetFolder

if data_split:
	train_data, val_data = random_split(data, data_split)
else:
	train-data = Subset(data, train_indices_from_file)
	val_data-data = Subset(data, val_indices_from_file)
```

During training I could basically save the `state_dict` of the `LightningDataModule` in the current checkpoint, however, this will incur additional complexity in extracting the information about the data splits whenever I want to train another model with LightningCLI on the same splits.

For now, let's imagine the locally stored dataset does not change and I don't have to worry about the stored data.

My solution now is just overriding the `instantiate_classes` member function as follows:
```python
    class MyLightningCLI(LightningCLI):
        def instantiate_classes(self) -> None:
            """Instantiates the classes and sets their attributes."""
            self.config_init = self.parser.instantiate_classes(self.config)
            self.datamodule = self._get(self.config_init, "data")
            self.model = self._get(self.config_init, "model")
            self._add_configure_optimizers_method_to_model(self.subcommand)
            self.trainer = self.instantiate_trainer()

			# this is the new change
            self.datamodule.save_data_split_to_file(self.trainer.logger.log_dir)
```
where `datamodule.save_data_split_to_file` is just a member of the LightningDataModule class which implements the storing functionality. I also had to move the creation of the train and val data splits from `setup()` to the constructor of this class because otherwise I wouldn't have the information about the splits generated when calling `MyLightningCLI.instantiate_classes`

If we could add another hook called `after_instantiate_classes` called after `LightningCLI.instantiate_classes` then one wouldn't have to override `instantiate_classes` and the code would look more "lightning-ish".

Thank you

### Pitch

Add `after_instantiate_classes` hook to be called after `instantiate_classes` in LightningCLI.

**Comment by awaelchli** [comment]
@tchesler I think your use case can easily be realized by moving your code to the `prepare_data` hook in the datamodule. Create your split files there and save them. Then in the `setup()` or `*_dataloader()` methods read these splits and instantiate your dataloaders. Can you let us know if that works?

**Comment by tchesler** [comment]
@awaelchli Will I be able to access the current logging directory in you scenario such that I can save the split for example int `lightning_logs/version_xxx/train_split.csv`?

**Comment by tchesler** [comment]
This works. Thank you, I didn't think that the `DataModule` object holds a reference to the current trainer. The only thing that is still a bit "hacky" is the fact that one needs to create the logging directory in `prepare_data` which can lead to some problems, I think, when running in distributed mode. The `SaveConfigCallback` makes sure to create this folder only on RANK 0. 

But, I tried your code without the overridden `prepare_data` function and it still works. I guess the `train_dataloader` function is called after the Logging has been properly set up. 

There still remains the question, whether a `after_instantiate_classes` member function in LightningCLI could be implemented in the future.

**Comment by awaelchli** [comment]
The fact that the directory is not created yet is an implementation detail and unrelated to the CLI. The prepare_data hook only gets executed on local rank 0 so it is save to write files or download data. This is what the hook was made for.

> But, I tried your code without the overridden prepare_data function and it still works.

I don't believe you 😅  If you remove the prepare_data hook from my example, you will get the assertion error:
```
  File "/Users/adrian/repositories/lightning/examples/pytorch/bug_report/bug_report_model.py", line 32, in train_dataloader
    assert Path(self.trainer.log_dir, "train-test-split.txt").exists()
AssertionError
```

> There still remains the question, whether a after_instantiate_classes member function in LightningCLI could be implemented in the future.

True, if there is a strong need for this backed by use cases, we will consider adding new APIs to Lightning. In your particular case however I think the Lightning way of using the datamodule hooks is the intended way. Because following this practice will not only make the code work with the CLI.

**Comment by tchesler** [comment]
> I don't believe you 😅 If you remove the prepare_data hook from my example, you will get the assertion error:
Ah, sorry about that. Ofc `train-test-split.txt` won't exist in your example if I comment out the `prepare_data` function. However the `trainer.log_dir` will exist at that point in time. That's what I meant. I basically removed the `prepare_data` code, and I am directly saving the file in `train_dataloader`:
```
def train_dataloader(self):
        pd.DataFrame(self.data.samples).to_csv(os.path.join(self.trainer.log_dir, 'data_samples.csv'))
        pd.DataFrame(self.train_data.indices).to_csv(os.path.join(self.trainer.log_dir, 'train_indices.csv'))
```
where `self.data` is my underlying `torch.Dataset` and `self.train_data` is the `Subset` generated with `random_split`.

**Comment by awaelchli** [comment]
Sounds good. In this case I suggest to add this:
```py
if self.trainer.global_rank == 0:
        pd.DataFrame(self.data.samples).to_csv(os.path.join(self.trainer.log_dir, 'data_samples.csv'))
        pd.DataFrame(self.train_data.indices).to_csv(os.path.join(self.trainer.log_dir, 'train_indices.csv'))
```
